container/src/lib.rs

```diff
@@ -86,20 +87,23 @@ impl Container {
     pub fn run(&self) -> Result<()> {
         let mounts = self.mounts.clone();
         let code = unsafe {
-            unshare::Command::from(&self.command)
+            let mut child = match unshare::Command::from(&self.command)
                 .chroot_dir(&self.rootfs)
                 .unshare(&*self.namespaces.get())
                 .pre_exec(move || Mounts::apply(&mounts))
                 .envs(self.environment.get())
                 .spawn()
-                .map_err(Error::ContainerSpawnCommand)?
-                .wait()
-                .map_err(Error::ContainerWaitCommand)?
-                .code()
+            {
+                Ok(child) => child,
+                Err(_) => {
+                    let _ = &self.mounts.cleanup(self.rootfs.clone())?;
+                    exit(2);
```

Comment:
You should not `exit`here but propagate the cleanup error by just `return self.mounts.cleanup(self.rootfs.clone());`

Reply:
Ok, I changed that but I have a question. 
When I tried to write `self.mounts.cleanup(self.rootfs.clone())` instead of `return self.mounts.cleanup(self.rootfs.clone())`, I got an error `expected struct unshare::Child, found enum std::result::Result`.
Is it not the same to write `lastline` without semicolon and to write `return lastline;` ?

Reply:
>  Is it not the same to write `lastline` without semicolon and to write `return lastline;` ?

The former is what the current scope (what's between `{}`) while the latter is a return from the root function scope (return from function).

In your example, you get an error because the scope is the `match`, and you bind a variable to the return value of the scope. If on the `Ok(child)` path your scope returns a `unshare::Child`, the `Err()` should also return the same type. If you call `return` from the `Err()` branch, then what you should return is of the same type as the function/method return value.
I hope it makes sense.

Reply:
Yes it's clear, thank you for your response !